### PR TITLE
nlohmann_json: Add support for Conan V2

### PR DESCRIPTION
**Pull request by jwillikers**
Specify library name and version:  **nlohmann_json/***

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/nlohmann_json/all/conanfile.py [hunk 1]
```diff
@@ -1,32 +1,37 @@
...
-        return "source_subfolder"
+    def layout(self):
+        basic_layout(self, src_folder="src")
 
     def package_id(self):
-        self.info.header_only()
+        self.info.clear()
+
+    def generate(self):
+        pass
```
Comment (danimtb) [inline recipes/nlohmann_json/all/conanfile.py:27]: probably this is not needed

**Review by danimtb — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

File: recipes/nlohmann_json/all/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/nlohmann_json/all/conanfile.py:27]: @danimtb  You would think so wouldn't you? It was tricky to figure out, but there was an error without this. No generators are specified, but Conan seems to default to one or something and fail? See my error output when I remove those two lines here:

```
conan create . nlohmann_json/3.11.2@ -pr:b linux/gcc/release -pr:h linux/gcc/debu
g -u -b outdated -b cascade
[HOOK - conan-center.py] pre_export(): [DEPRECATED GLOBAL CPPSTD (KB-H001)] OK
[HOOK - conan-center.py] pre_export(): [REFERENCE LOWERCASE (KB-H002)] OK
[HOOK - conan-center.py] pre_export(): [RECIPE METADATA (KB-H003)] OK
[HOOK - conan-center.py] pre_export(): [HEADER_ONLY, NO COPY SOURCE (KB-H005)] OK
[HOOK - conan-center.py] pre_export(): [FPIC OPTION (KB-H006)] OK
[HOOK - conan-center.py] pre_export(): [VERSION RANGES (KB-H008)] OK
[HOOK - conan-center.py] pre_export(): [RECIPE FOLDER SIZE (KB-H009)] Total recipe size: 9.5947265625 KB
[HOOK - conan-center.py] pre_export(): [RECIPE FOLDER SIZE (KB-H009)] OK
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] exports: None
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] exports: None
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE FOLDER (KB-H024)] OK
[HOOK - conan-center.py] pre_export(): [META LINES (KB-H025)] OK
[HOOK - conan-center.py] pre_export(): [CONAN CENTER INDEX URL (KB-H027)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE MINIMUM VERSION (KB-H028)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - RUN ENVIRONMENT (KB-H029)] OK
[HOOK - conan-center.py] pre_export(): [SYSTEM REQUIREMENTS (KB-H032)] OK
[HOOK - conan-center.py] pre_export(): [CONANDATA.YML FORMAT (KB-H030)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - NO IMPORTS() (KB-H034)] OK
[HOOK - conan-center.py] pre_export(): [NO AUTHOR (KB-H037)] OK
[HOOK - conan-center.py] pre_export(): [NOT ALLOWED ATTRIBUTES (KB-H039)] OK
[HOOK - conan-center.py] pre_export(): [NO TARGET NAME (KB-H040)] OK
[HOOK - conan-center.py] pre_export(): [NO FINAL ENDLINE (KB-H041)] OK
[HOOK - conan-center.py] pre_export(): [NO REQUIRES.ADD() (KB-H044)] OK
[HOOK - conan-center.py] pre_export(): [DELETE OPTIONS (KB-H045)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE VERBOSE MAKEFILE (KB-H046)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE VERSION REQUIRED (KB-H048)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE WINDOWS EXPORT ALL SYMBOLS (KB-H049)] OK
[HOOK - conan-center.py] pre_export(): [DEFAULT OPTIONS AS DICTIONARY (KB-H051)] OK
[HOOK - conan-center.py] pre_export(): [CONFIG.YML HAS NEW VERSION (KB-H052)] OK
[HOOK - conan-center.py] pre_export(): [PRIVATE IMPORTS (KB-H053)] OK
[HOOK - conan-center.py] pre_export(): [SINGLE REQUIRES (KB-H055)] OK
[HOOK - conan-center.py] pre_export(): [TOOLS RENAME (KB-H057)] OK
[HOOK - conan-center.py] pre_export(): [ILLEGAL CHARACTERS (KB-H058)] OK
[HOOK - conan-center.py] pre_export(): [CLASS NAME (KB-H059)] OK
[HOOK - conan-center.py] pre_export(): [NO CRLF (KB-H060)] OK
[HOOK - conan-center.py] pre_export(): [NO BUILD SYSTEM FUNCTIONS (KB-H061)] OK
[HOOK - conan-center.py] pre_export(): [TOOLS CROSS BUILDING (KB-H062)] OK
[HOOK - conan-center.py] pre_export(): [INVALID TOPICS (KB-H064)] OK
[HOOK - conan-center.py] pre_export(): [NO REQUIRED_CONAN_VERSION (KB-H065)] OK
[HOOK - conan-center.py] pre_export(): [TEST_TYPE MANAGEMENT (KB-H068)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - NO DEFAULT OPTIONS (KB-H069)] OK
[HOOK - conan-center.py] pre_export(): [MANDATORY SETTINGS (KB-H070)] OK
Exporting package recipe
nlohmann_json/3.11.2 exports: File 'conandata.yml' found. Exporting it...
nlohmann_json/3.11.2 exports: Copied 1 '.yml' file: conandata.yml
[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] Saving conandata.yml: {'sources': {'3.11.2': {'url': 'https://github.com/nlohmann/json/archive/v3.11.2.tar.gz', 'sha256': 'd69f9deb6a75e2580465c6c4c5111b89c4dc2fa94e3a85fcd2ffcd9a143d9273'}}}
[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] New conandata.yml contents: sources:
  3.11.2:
    sha256: d69f9deb6a75e2580465c6c4c5111b89c4dc2fa94e3a85fcd2ffcd9a143d9273
    url: https://github.com/nlohmann/json/archive/v3.11.2.tar.gz

[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] OK
[HOOK - conan-center.py] post_export(): [DEFAULT SHARED OPTION VALUE (KB-H050)] OK
[HOOK - conan-center.py] post_export(): [LICENSE PUBLIC DOMAIN (KB-H056)] OK
nlohmann_json/3.11.2: A new conanfile.py version was exported
nlohmann_json/3.11.2: Folder: /var/home/jordan/.conan/data/nlohmann_json/3.11.2/_/_/export
nlohmann_json/3.11.2: Using the exported files summary hash as the recipe revision: 33798f32eda5d87713e03a065f4cc05f 
nlohmann_json/3.11.2: Exported revision: 33798f32eda5d87713e03a065f4cc05f
Configuration (profile_host):
[settings]
arch=x86_64
build_type=Debug
compiler=gcc
compiler.cppstd=17
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
[options]
[build_requires]
[env]
[conf]
tools.env.virtualenv:auto_use=True
tools.env.virtualenv:powershell=False
tools.cmake.cmake_layout:build_folder_vars=['settings.arch', 'settings.compiler', 'options.shared']
tools.cmake.cmaketoolchain:find_package_prefer_config=True
tools.cmake.cmaketoolchain:generator=Ninja
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True

Configuration (profile_build):
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=17
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
[options]
[build_requires]
[env]
[conf]
tools.env.virtualenv:auto_use=True
tools.env.virtualenv:powershell=False
tools.cmake.cmake_layout:build_folder_vars=['settings.arch', 'settings.compiler', 'options.shared']
tools.cmake.cmaketoolchain:find_package_prefer_config=True
tools.cmake.cmaketoolchain:generator=Ninja
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True

nlohmann_json/3.11.2: WARN: The package nlohmann_json/3.11.2:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 doesn't belong to the installed recipe revision, removing folder
nlohmann_json/3.11.2 (test package): Installing package
Requirements
    nlohmann_json/3.11.2 from local cache - No remote
Packages
    nlohmann_json/3.11.2:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Build

Installing (downloading, building) binaries...
nlohmann_json/3.11.2: WARN: Build folder is dirty, removing it: /var/home/jordan/.conan/data/nlohmann_json/3.11.2/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
[HOOK - conan-center.py] pre_source(): [IMMUTABLE SOURCES (KB-H010)] OK
nlohmann_json/3.11.2: Configuring sources in /var/home/jordan/.conan/data/nlohmann_json/3.11.2/_/_/source/src
Downloading v3.11.2.tar.gz completed [7907.88k]                                          nlohmann_json/3.11.2: n_json/3.11.2: 
nlohmann_json/3.11.2: 
[HOOK - conan-center.py] post_source(): [LIBCXX MANAGEMENT (KB-H011)] OK
[HOOK - conan-center.py] post_source(): [CPPSTD MANAGEMENT (KB-H022)] OK
[HOOK - conan-center.py] post_source(): [SHORT_PATHS USAGE (KB-H066)] OK
nlohmann_json/3.11.2: Building your package in /var/home/jordan/.conan/data/nlohmann_json/3.11.2/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
nlohmann_json/3.11.2: Generator txt created conanbuildinfo.txt
Traceback (most recent call last):
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/command.py", line 2216, in run
    method(args[0][1:])
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/command.py", line 380, in create
    info = self._conan.create(args.path, name=name, version=version, user=user,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/conan_api.py", line 93, in wrapper
    return f(api, *args, **kwargs)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/conan_api.py", line 388, in create
    create(self.app, ref, graph_info, remotes, update, build_modes,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/cmd/create.py", line 72, in create
    install_build_and_test(app, test_conanfile_path,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/cmd/test.py", line 35, in install_build_and_test
    deps_install(app=app,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/manager.py", line 83, in deps_install
    installer.install(deps_graph, remotes, build_modes, update, profile_host, profile_build,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 317, in install
    self._build(nodes_by_level, keep_build, root_node, profile_host, profile_build,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 464, in _build
    self._handle_node_cache(node, keep_build, processed_package_refs, remotes)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 557, in _handle_node_cache
    pref = self._build_package(node, output, keep_build, remotes)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 600, in _build_package
    pref = builder.build_package(node, keep_build, self._recorder, remotes)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 227, in build_package
    self._build(conanfile, pref)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 138, in _build
    write_toolchain(conanfile, conanfile.generators_folder, self._output)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/generators/__init__.py", line 257, in write_toolchain
    with chdir(path):
  File "/usr/lib64/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/tools/files.py", line 27, in chdir
    os.chdir(newdir)
FileNotFoundError: [Errno 2] No such file or directory: '/var/home/jordan/.conan/data/nlohmann_json/3.11.2/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/build-debug/conan'

ERROR: [Errno 2] No such file or directory: '/var/home/jordan/.conan/data/nlohmann_json/3.11.2/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/build-debug/conan'
```

Comment (uilianries) [reply recipes/nlohmann_json/all/conanfile.py:27]: WTF?!

Comment (jwillikers) [reply recipes/nlohmann_json/all/conanfile.py:27]: Conan version 1.51.3.

Comment (jwillikers) [reply recipes/nlohmann_json/all/conanfile.py:27]: Seems to be related to adding a `layout` method for a header-only library which doesn't use the `generate` method or specify any `generators`.

**Pull request merged**